templates/lib/prism/node.rb.erb

```diff
@@ -96,6 +96,39 @@ module Prism
     def type
       raise NoMethodError, "undefined method `type' for #{inspect}"
     end
+
+    # Value equality.
+    def ==(other)
+      return false unless other.is_a?(Node) && other.type == type
+
+      deconstructed_self = deconstruct_keys(nil)
```

Comment:
Ahh, I really don't want to tie our pattern matching implementation to equality. We could easily add things to pattern matching that would could break this (like the ability to take out the `value` of an integer node, or the options of a regexp). This adds a dependency that we don't want to have.

This also doesn't account for source, which we definitely want to track. Nodes are tied to the source file they came from, but this doesn't account for that.

Reply:
I agree it makes sense to track source if we track location. Will look at making this independent from pattern matching. For context, I took inspiration from the node equality test helper - so that dependency is already there internally, might make sense fixing that. Maybe a good path forward becomes obvious when I find a better solution for this PR.

Reply:
> the ability to take out the value of an integer node

That would actually just make things more consistent. Why _isn't_ the value included in the deconstructed version of a node?

Reply:
Oh, also - this is a recursive comparison, but the source would be the same for the whole tree, right? We wouldn't want to compare the source for each node of the tree.

Reply:
> We wouldn't want to compare the source for each node of the tree.

If it's by identity it wouldn't matter much performance-wise. I wouldn't imagine to compare Source bytes, that sounds too expensive.
Which makes me think, maybe Node#== should be identity, just like it already is. Is there a use case for a `Node#==` which would compare locations & source & fields? Wouldn't the 2 nodes need to be the same to be == then?
One case I could think of is if using Prism::Compiler and transforming the tree, in that cases unmodified nodes could be ==. But I'm not sure there is any point to call == on these subtrees.

Based on these thoughts, my suggestion would be to focus on `===` (and that would ignore locations and so the source too, per the discussion on the issue) and leave `==` as identity until there a use case to do something different.

Reply:
Yeah, I think that makes sense to me. I really want to know whether two different sources parse into equivalent ASTs.